Cotracker (#172)
creatis-myriad/creatis-myriad.github.io: collections/_posts/2024-04-16-cotracker.md

---
layout: review
title: "CoTracker: It is Better to Track Together"
tags: motion estimation, transformer
author: "Olivier Bernard"
cite:
    authors: "Nikita Karaev, Ignacio Rocco, Benjamin Graham, Natalia Neverova, Andrea Vedaldi, Christian Rupprecht"
    title: "CoTracker: It is Better to Track Together"
    venue: "Arxiv 2023"
pdf: "https://arxiv.org/pdf/2307.07635.pdf"
---

# Notes

* The [official webpage](https://co-tracker.github.io/) with paper/code/demo 

&nbsp;

# Highlights

* Transformer-based model that tracks dense points in a frame jointly across a video sequence
* Joint tracking results in a significantly higher tracking accuracy and robustness

> * Can track arbitrary points, selcted at any spatial location and time in the video 

Reviewer: ```suggestion
* Can track arbitrary points, selected at any spatial location and time in the video 
```

* Operates causally on short windows (suitable for online tasks)
* Trained by unrolling the windows across longer video sequences, which improves long-term tracking

* Introduce the concept of virtual tracks to track more than 70k points jointly and simultaneously

> * Introduce the concept of support points to improve performance by jointly track additional points to reinforce contextualization

Reviewer: ```suggestion
* Introduce the concept of support points to improve performance by jointly tracking additional points to reinforce contextualization
```

$$\left\{
\begin{aligned}
    & P_t^i = \left(x_t^i,y_t^i\right) \in \mathbb{R}^2 \\
    & i \in \{1, \cdots N \} \\
    & t \in \{t_i,\cdots,T \} \\
    & t_i \in \{1, \cdots, N\} \quad \text{is the time when the track starts}
\end{aligned}
\right.$$

![](/collections/images/cotracker/context-formulation.jpg)

&nbsp;

* CoTracker also estimates the visibility flag $$\nu_i \in \{0,1\}$$ which tells if a point is visible or occluded in a given frame

&nbsp;

## Information extaction

**_1) Image features_**

* $$\phi(I_t) \in \mathbb{R}^{d \times \frac{H}{4} \times \frac{W}{4}}$$ are _d_-dimensional appearance features that are extracted from each video frame using a CNN which is trained end-to-end. 

* The resolution is reduced by a factor of $$4$$ for efficiency

* $$\phi_s(I_t) \in \mathbb{R}^{d \times \frac{H}{4 \cdot 2^{s-1}} \times \frac{W}{4 \cdot 2^{s-1}}}$$ are scaled versions of the appearance features used to compute correlation features

* $$s \in \{ 1, \cdots, S \}$$ with $$S=4$$ in their implementation

![](/collections/images/cotracker/image-features.jpg)

&nbsp;

**_2) Appearance features_**

* $$Q_t^i \in \mathbb{R}^d$$ are appearance features of the **_tracks_** 

> * They are time dependent to accomodate changes in the track appearance

Reviewer: ```suggestion
* They are time dependent to accommodate changes in the track appearance
```

* $$\eta'(\cdot)$$ is a positional encoding of the start position $$\hat{P}_1^i$$ and for time $$t$$, with appropriate dimensions

* The estimates $$\hat{P}$$, $$\hat{\nu}$$ and $$Q$$ are initialized by broadcasting the initial values $$P_{t_i}^{i}$$ (the location of query point), $$1$$ (meaning visible) and $$\phi\left(I_t\right)[P_{t_i}^{i}/4]$$ (the appearance of the query point) to all time $$t \in \{1,\cdots,T\}$$

![](/collections/images/cotracker/cotracker-initialization.jpg)

&nbsp;

**_Output tokens_**

* The output tokens $$O\left( \Delta \hat{P}, \Delta Q \right)$$ contains updates for location and appareance, i.e. $$O_t^i = \left( \Delta \hat{P}_t^i, \Delta Q_t^i \right)$$

&nbsp;

**_Iterated transformer applications_**

* The track estimates are progressively refined using an iterative process $$m \in \{0,\cdots,M\}$$

* $$m=0$$ denotes initialization

* Each update computes

$$\left(\Delta \hat{P}, \Delta Q\right) = \Psi\left( G\left( \hat{P}^{(m)}, \, \hat{\nu}^{(0)}, \, Q^{(m)} \right)\right)$$

$$\left\{
\begin{aligned}
    & \hat{P}^{(m+1)} = \hat{P}^{(m)} + \Delta \hat{P} \\
    & Q^{(m+1)} = Q^{(m)} + \Delta Q
\end{aligned}
\right.$$

> * The visibility mask $$\hat{\nu}$$ is not updated by the transformer but once after the last $$M$$ applications of the tansformer usign the appareance features as $$\hat{\nu}^{(M)} = \sigma\left(W Q^{(M)}\right)$$

Reviewer: ```suggestion
* The visibility mask $$\hat{\nu}$$ is not updated by the transformer but once after the last $$M$$ applications of the transformer using the appearance features as $$\hat{\nu}^{(M)} = \sigma\left(W Q^{(M)}\right)$$
```

* The transformer $$\Psi$$ interleaves attention operators that operate across the _time_ and _point track_ dimensions, so to reduce the complexity from $$O( N^2 T^2)$$ to $$O( N^2 + T^2 )$$ 

* Virtual tracks concept is introduced to even reduce the complexity when $$N$$ is large 

* Virtual tracks tokens are simply added as additional tokens to the input grid with a fixed, learnable initialization and removed at the output

* $$K$$ virtual tracks tokens reduce the complexity to $$O( NK + K^2 + T^2 )$$

![](/collections/images/cotracker/cotracker-algorithm.jpg)

&nbsp;

## Training process

> * Use of sliding window concept to handle with long videos of lenght $$T'>T$$, where $$T$$ is the maximum window length supported by the architecture

Reviewer: ```suggestion
* Use of sliding window concept to handle  long videos of length $$T'>T$$, where $$T$$ is the maximum window length supported by the architecture
```

&nbsp;

![](/collections/images/cotracker/davis_car.gif)
![](/collections/images/cotracker/davis_libby.gif)
![](/collections/images/cotracker/davis_soapbox.gif)

<p style="text-align: center;font-style:italic">Figure 1 - State of the art comparison. First column: PIPs, second column: RAFT, third column: TAPIR, last column: CoTracker.</p>

&nbsp;

## Is unrolled training important ?

<div style="text-align:center">
<img src="/collections/images/cotracker/results-unrolled-training.jpg" width=350></div>
<p style="text-align: center;font-style:italic">Table 3. Unrolled training. CoTracker is built for sliding window predictions. Using them during training is important.</p>

&nbsp;

## Do virtual tracks help CoTracker to scale ?

<div style="text-align:center">
<img src="/collections/images/cotracker/results-virtual-tracks.jpg" width=350></div>
<p style="text-align: center;font-style:italic">Table 4. The virtual tracks allow CoTracker to scale. We report the maximum number of tracks that can fit on a 80 GB GPU.</p>

&nbsp;

# Conclusions

* Joint tracking of points improves results
* CoTracker is among the current best performing method for point tracking

> * Introduce several interesting concepts, such as occuracy flags and support points

Reviewer: occlusion flags?
